abci/tests/common.rs

```diff
@@ -0,0 +1,161 @@
+        };
+
+        td.id = td.runtime.block_on(td.start())?;
+
+        // Handle ctrl+c
+        let id = td.id.clone();
+        let docker = td.docker.clone();
+        // let runtime = &td.runtime;
+        let _ctrlc = td.runtime.spawn(async move {
+            tokio::signal::ctrl_c().await.unwrap();
+            debug!("Received ctrl+c, removing Tenderdash container");
+            timeout(Duration::from_secs(15), Self::stop(id, &docker))
+                .await
+                .expect("timeout removing tenderdash container");
+            std::process::exit(1);
+        });
+
+        Ok(td)
+    }
+
+    async fn connect() -> Result<Docker, Error> {
+        debug!("Connecting to Docker server");
+        let docker = Docker::connect_with_socket("/var/run/docker.sock", 120, API_DEFAULT_VERSION)?;
+
+        let info = docker.info().await?;
+        debug!(
+            "Connected to Docker server version {}",
+            info.server_version.unwrap()
+        );
+
+        Ok(docker)
+    }
+
+    async fn image_pull(&self) -> Result<(), Error> {
+        debug!("Fetching image {}", self.image);
+        let image_responses = self.docker.create_image(
+            Some(bollard::image::CreateImageOptions {
+                from_image: self.image.clone(),
+                ..Default::default()
+            }),
+            None,
+            None,
+        );
+
+        let mut image_responses = image_responses
+            .collect::<Vec<Result<CreateImageInfo, bollard::errors::Error>>>()
+            .await;
+
+        let image_response = image_responses.pop().unwrap()?;
+        debug!("Image fetch status: {}", image_response.status.unwrap());
+        Ok(())
+    }
+
+    async fn create_container(&self) -> Result<String, Error> {
+        debug!("Creating container");
+        let container_config = Config {
+            image: Some(self.image.clone()),
+            env: Some(vec![String::from("PROXY_APP=unix:\\/\\/\\/abci.sock")]),
+            host_config: Some(HostConfig {
+                binds: Some(vec![String::from("/tmp/socket:/abci.sock")]),
+                ..Default::default()
+            }),
+            ..Default::default()
+        };
+
+        let id = self
+            .docker
+            .create_container::<String, String>(
+                Some(bollard::container::CreateContainerOptions {
+                    name: String::from("tenderdash"),
+                    ..Default::default()
+                }),
+                container_config,
+            )
+            .await?
+            .id;
+
+        debug!("Starting container");
+        self.docker
+            .start_container::<String>(
+                &id,
+                Some(bollard::container::StartContainerOptions {
+                    ..Default::default()
+                }),
+            )
+            .await?;
+
+        Ok(id)
+    }
+    /// Start Tenderdash in Docker
+    async fn start(&self) -> Result<String, Error> {
```

Comment:
A common approach in Rust is to have initialized value to mean "acquired resource", like an instance of `File` means that you already have a file handle, so it's not a `file.open()`. What I propose here is to try to do the same thing here, like a created instance of `TenderdashDocker` means a running container, so no need for a separate `.start` method

Reply:
Could you check with some debug print if `Drop` will be called on ctrl-c ? I think it should;

About `start`, I can't see if is ever called

Reply:
ah, got it, thanks